docs/compactor.md

```diff
@@ -0,0 +1,86 @@
+    * Incremental-Snapshot-revision-7
+    * Incremental-Snapshot-revision-13
+    * Incremental-Snapshot-revision-15 (It containes command upto revision 18)
+    * Full-Snapshot-revision-18 (Compacted)
+    * Incremental-Snapshot-revision-19
+    * Incremental-Snapshot-revision-23
+    * Incremental-Snapshot-revision-28
+    * Incremental-Snapshot-revision-33 (It containes command upto revision 38)
+    * Full-Snapshot-revision-38 (Compacted)
+    * Incremental-Snapshot-revision-39
+    * Incremental-Snapshot-revision-42
+    * Incremental-Snapshot-revision-43
+    * Incremental-Snapshot-revision-49
+
+3. A meta-data file might need to be added in the Backup directory to mark which compacted snapshot is ready to be used for restoration. Restorer would study the meta-data file and pickup the ready full snapshot from there and apply all the delta snapshots having revision number greater than that. So, with the backup directory as above, if the content of metadata file is like following:
+
+* Full-Snapshot-revision-1
+* Full-Snapshot-revision-18
+
+Then, during restoration Full-Snapshot-revision-18 and all the delta snaphots from revision number 19 to 49 will be applied. **Restorer would restore from latest ready full snapshot and all subsequent delta snapshots.** As Full-Snapshot-revision-38 is not present in the metadata file, it would not be considered as ready yet.
+
+Another approach to make sure that restorer only picks up snapshots which are ready, full snapshots can be named with extension `.part` until the snapshot is ready. When the snapshot is ready, `.part` extension can be removed from it's name. Restorer would avoid any file having extension as `.part`. 
+
+So, if Full-Snapshot-revision-38 is not ready in the above example, it can be named as `Full-Snapshot-revision-38.part`. Restorer will apply Full-Snapshot-revision-18 and all the delta snaphots from revision number 19 to 49. **Restorer would restore from latest ready full snapshot and all subsequent delta snapshots.**
+
+4. Keep provisions for backward compatibility so that ETCDBR can work with both old and new type of backup store. We may want to introduce a metadata file in object store for new type of backup structure. If ETCDBR encounter metadata file suggesting a backup store contains new type of directory structure, it switches to compact mode.
+5. **Garbage collection** : Garbage collection is very much important for this compaction story. As the restorer have to look for latest compacted full snapshots from among the numerous full snapshots already present in a directory, it is essential to reduce the number of search space. Garbage collecter should work based on some policy and regularly clean up the directory from old unwanted snapshots. 
+
+**Note:** ETCD Backup Restore needs to take one full snapshot just after starting. Compactor will work only if there is any full snapshot already present in the store. It is not limitation but a design choice
```

Comment:
```suggestion
#### Backward Compatibility

1. **Restoration** : The changes to handle the newly proposed backup directory structure must be backward compatible with older structures at least for restoration because we need have to restore from backups in the older structure. This includes the support for restoring from a backup without a metadata file if that is used in the actual implementation.
2. **Backup** : For new snapshots (even on a backup containing the older structure), the new structure may be used. The new structure must be setup automatically including creating the base full snapshot (and also the creation of a metadata file if necessary).
3. **Garbage collection** : The existing functionality of garbage collection of snapshots (full and incremental) that fall out of the backup retention policy must be compatible with both old and new backup folder structure. I.e. the snapshots in the older backup structure must be retained in their own structure and the snapshots in the proposed backup structure should be retained in the proposed structure. Once all the snapshots in the older backup structure go out of the retention policy and are garbage collected, we can think of removing the support for older backup folder structure.  

**Note:** ETCD Backup Restore needs to take one full snapshot just after starting. Compactor will work only if there is any full snapshot already present in the store. It is not limitation but a design choice
```